registryoverwrite.go

```diff
@@ -0,0 +1,181 @@
+package skynet
+
+// registry.go contains helper functions for working with the Skynet registry
+// in go.
+
+import (
+	"bytes"
+	"encoding/hex"
+	"encoding/json"
+	"io/ioutil"
+	"net/url"
+
+	"gitlab.com/NebulousLabs/errors"
+	"gitlab.com/SkynetLabs/skyd/node/api"
+	"go.sia.tech/siad/crypto"
+	"go.sia.tech/siad/modules"
+	"go.sia.tech/siad/types"
+)
+
+// ErrRegistryEntryNotFound is returned when the registry entry was not found.
+var ErrRegistryEntryNotFound = errors.New("registry entry was not found")
+
+// ReadRegistry will read a registry value from Skynet, collecting both the
+// data and the latest revision number.
+//
+// NOTE: Because this data is signed, the portal cannot provide a false
+// response. The portal can however provide an out-of-date response and present
+// it as the latest information. If we want to discourage this, we will need
+// the portal to respond with some of the host responses, showing that the
+// request was made recently and that the host has signed off that this is the
+// latest data. At the very least, this would require the portal to be
+// colluding with all the hosts it is providing as evidence for it to be able
+// to lie to the user. We could take this even a step further by having a list
+// of hosts that we want the portal to provide responses from, where we expect
+// at least some percentage (such as 2/3rds) of the hosts to have responded
+// successfully. This even further reduces any chance the portal has to lie to
+// us.
+func (sc *SkynetClient) ReadRegistry(spk types.SiaPublicKey, dataKey crypto.Hash) (data []byte, revisionNumber uint64, err error) {
+	// Fetch the current revision number.
+	values := url.Values{}
+	values.Set("publickey", spk.String())
+	values.Set("datakey", dataKey.String())
+	resp, err := sc.executeRequest(
+		requestOptions{
+			Options: DefaultOptions("/skynet/registry"),
+			method:  "GET",
+			query:   values,
+		},
+	)
+	if err != nil {
+		return nil, 0, errors.AddContext(err, "unable to execute request")
+	}
+	defer resp.Body.Close()
+
+	// A 404 is not an error, it just means there is no data.
+	if resp.StatusCode == 404 {
+		return nil, 0, ErrRegistryEntryNotFound
+	}
+
+	// Parse the response.
+	if resp.ContentLength > 10e6 {
```

Comment:
wouldn't you usually use a limit reader? I believe this header can be overwritten with whatever value the portal wants it to be.

Reply:
Agreed. Blindly using `ioutil.ReadAll` won't be safe because we can't necessarily trust the content length here. I'd wrap the `resp` in a `io.LimitReader` first and then use a `dec := json.NewDecoder(limitReader)` since that's more efficient than `ioutil.ReadAll`.